pkg/virt-launcher/virtwrap/device/hostdevice/sriov/hostdev.go

```diff
@@ -34,7 +40,49 @@ import (
 
 func CreateHostDevices(vmi *v1.VirtualMachineInstance) ([]api.HostDevice, error) {
 	SRIOVInterfaces := vmispec.FilterSRIOVInterfaces(vmi.Spec.Domain.Devices.Interfaces)
-	return CreateHostDevicesFromIfacesAndPool(SRIOVInterfaces, NewPCIAddressPool(SRIOVInterfaces))
+	netStatusPath := path.Join(sriov.MountPath, sriov.VolumePath)
+	pciAddressPoolWithNetworkStatus, err := newPCIAddressPoolWithNetworkStatusFromFile(netStatusPath)
+	if err != nil {
+		return nil, err
+	}
+	if pciAddressPoolWithNetworkStatus.Len() == 0 {
+		log.Log.Object(vmi).Warningf("found no SR-IOV networks to PCI-Address mapping. fall back to resource address pool")
+		return CreateHostDevicesFromIfacesAndPool(SRIOVInterfaces, NewPCIAddressPool(SRIOVInterfaces))
+	}
+
+	return CreateHostDevicesFromIfacesAndPool(SRIOVInterfaces, pciAddressPoolWithNetworkStatus)
+}
+
+// newPCIAddressPoolWithNetworkStatusFromFile polls the given file path until populated, then uses it to create the
+// PCI-Address Pool.
+// possible return values are:
+// - file populated - return PCI-Address Pool using the data in file.
+// - file empty post-polling (timeout) - return  err.
```

Comment:
Can you please explain what is the reason behind the decision to return an error just when the file is empty.

Reply:
Yes. 
First concern is that we want to keep same behavior like before (fallback to old mapping) in cases such as network-status annotation is not in the pod or the PCI data is not int the annotation (as in old Multus deployed or bug in their annotation). 
Other case when we want to "fallback" is when the volume does not exist (possible bug in downwardAPI) or we error reading it. 
In "fallback" scenario - we will return nil but the outputted `pciAddressPoolWithNetworkStatus.Len()` will be 0, thus causing the fallback to [occur](https://github.com/kubevirt/kubevirt/pull/8226/files/b0f2ba81f832e8ef1de0adf15535495aa093d91d..0b93379dfb2e0ee7f42c41d3fb17159cfed5b3cb#diff-4bf5d16bba47083655ec1792df984d6426e722cc13b1aceb5184218146b56f9cR48).

Second concern is that since downwardAPI sometimes takes time for the data to propagate to the volume inside the pod (sometimes is instant, sometimes 400ms by my measurements) . 
In order to not do a SyncVMI err ping pong until the volume is updated (unnecessarily expensive traffic wise), we introduced a short poll of 1 second and then return error so that the SyncVMI gRPC will re-enqueue (this will also propagate the syncError as an event).

Reply:
Thanks! I meant adding a comment to the code:)

Reply:
ah, :) 
Done